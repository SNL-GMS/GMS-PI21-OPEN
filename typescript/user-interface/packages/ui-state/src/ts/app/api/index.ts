export * from './data';
export * from './event-manager';
export * from './processing-configuration';
export * from './processing-station';
export * from './signal-enhancement-configuration';
export * from './soh-acei';
export * from './ssam-control';
export * from './station-definition';
export * from './system-event-gateway';
export * from './system-message-definition';
export * from './user-manager';
export * from './util';
export * from './workflow';
