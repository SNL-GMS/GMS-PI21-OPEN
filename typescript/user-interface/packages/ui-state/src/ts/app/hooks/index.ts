export * from './channel-segment-hooks';
export * from './data-acquisition-hooks';
export * from './event-manager-hooks';
export * from './filter-hooks';
export * from './keyboard-shortcuts-hooks';
export * from './operational-time-period-configuration-hooks';
export * from './processing-analyst-configuration-hooks';
export * from './processing-station-hooks';
export * from './query-util-hooks';
export * from './react-redux-hooks';
export * from './signal-detection-hooks';
export * from './station-definition-hooks';
export * from './system-event-gateway-hooks';
export * from './user-manager-hooks';
export * from './waveform-hooks';
