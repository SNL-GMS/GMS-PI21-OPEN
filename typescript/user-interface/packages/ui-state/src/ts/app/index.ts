export * from './api';
export * from './hooks';
export * from './query';
export * from './redux-provider';
export * from './state';
export * from './store';
export * from './subscription';
