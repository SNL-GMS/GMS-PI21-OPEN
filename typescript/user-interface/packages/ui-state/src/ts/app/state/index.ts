export * as AppActions from './actions';
export * from './analyst';
export * from './common';
export * from './data-acquisition';
export * from './events';
export * from './keypress';
export * from './map';
export * as AppOperations from './operations';
export * from './signal-detections';
export * from './station-properties-configuration';
export * from './system-message';
export * from './user-session';
export * from './waveform';
export * from './workflow';
