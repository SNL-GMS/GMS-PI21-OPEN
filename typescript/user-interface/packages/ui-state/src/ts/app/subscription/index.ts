import * as SohStatusSubscription from './soh-status-subscription-wrapper';
import * as Subscription from './subscription';
import * as SystemMessageSubscription from './system-message-subscription';

export * from './types';
export { Subscription };
export { SohStatusSubscription };
export { SystemMessageSubscription };
