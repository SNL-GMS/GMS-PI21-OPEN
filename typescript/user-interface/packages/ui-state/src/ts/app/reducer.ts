import { combineReducers } from '@reduxjs/toolkit';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { dataSlice } from './api/data/data-slice';
import { reducer as appReducer } from './state/reducer';

export const reducer = combineReducers({
  // application api (queries)
  [systemEventGatewayApiSlice.reducerPath]: systemEventGatewayApiSlice.reducer,
  [eventManagerApiSlice.reducerPath]: eventManagerApiSlice.reducer,
  [processingConfigurationApiSlice.reducerPath]: processingConfigurationApiSlice.reducer,
  [processingStationApiSlice.reducerPath]: processingStationApiSlice.reducer,
  [signalEnhancementConfigurationApiSlice.reducerPath]:
    signalEnhancementConfigurationApiSlice.reducer,
  [ssamControlApiSlice.reducerPath]: ssamControlApiSlice.reducer,
  [sohAceiApiSlice.reducerPath]: sohAceiApiSlice.reducer,
  [stationDefinitionSlice.reducerPath]: stationDefinitionSlice.reducer,
  [systemMessageDefinitionApiSlice.reducerPath]: systemMessageDefinitionApiSlice.reducer,
  [userManagerApiSlice.reducerPath]: userManagerApiSlice.reducer,
  [workflowApiSlice.reducerPath]: workflowApiSlice.reducer,
  // data state
  [dataSlice.name]: dataSlice.reducer,
  // application state
  app: appReducer
});
