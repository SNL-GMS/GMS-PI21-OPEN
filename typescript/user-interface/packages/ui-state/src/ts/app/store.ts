import { IS_NODE_ENV_DEVELOPMENT, IS_NODE_ENV_PRODUCTION, NODE_ENV } from '@gms/common-util';
import { getElectron, getElectronEnhancer, UILogger } from '@gms/ui-util';
import { configureStore } from '@reduxjs/toolkit';
import type { ThunkMiddleware } from '@reduxjs/toolkit/node_modules/redux-thunk';
import thunk from '@reduxjs/toolkit/node_modules/redux-thunk';
import { createLogger } from 'redux-logger';
import {
  createStateSyncMiddleware,
  initStateWithPrevTab,
  withReduxStateSync
} from 'redux-state-sync';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { reducer } from './reducer';
import { userSessionActions } from './state';
import { waveformSlice } from './state/waveform';
import {
  establishWsConnection,
  registerConnectionStatusCallback
} from './subscription/subscription';

const DISABLE_REDUX_STATE_SYNC = process.env.GMS_DISABLE_REDUX_STATE_SYNC === 'true';

const logger = UILogger.create('GMS_LOG_REDUX_STORE', process.env.GMS_LOG_REDUX_STORE || 'info');

const electron = getElectron();
const electronEnhancer = electron ? getElectronEnhancer() : undefined;

if (electron && electronEnhancer) {
  logger.info('Configuring Redux store for Electron');
} else {
  logger.info('Configuring Redux store for browser');
}

logger.info(
  `Configuring Redux store with the following properties: ` +
    `NODE_ENV:${NODE_ENV} ` +
    `GMS_ENABLE_REDUX_LOGGER:${process.env.GMS_ENABLE_REDUX_LOGGER} ` +
    `GMS_DISABLE_REDUX_IMMUTABLE_CHECK:${process.env.GMS_DISABLE_REDUX_IMMUTABLE_CHECK} ` +
    `GMS_DISABLE_REDUX_SERIALIZABLE_CHECK:${process.env.GMS_DISABLE_REDUX_SERIALIZABLE_CHECK} ` +
    `DISABLE_REDUX_STATE_SYNC:${DISABLE_REDUX_STATE_SYNC} `
);

const thunkMiddleware: ThunkMiddleware = thunk;

const buildStore = () => {
  const store = configureStore({
    reducer: !DISABLE_REDUX_STATE_SYNC ? withReduxStateSync(reducer) : reducer,
    devTools:
      !IS_NODE_ENV_PRODUCTION &&
      process.env.GMS_DISABLE_REDUX_DEV_TOOLS?.toLocaleLowerCase() !== 'true'
        ? {
            trace: true,
            actionsBlacklist: [
              waveformSlice.actions.incrementLoadingTotal.name,
              waveformSlice.actions.incrementLoadingCompleted.name
            ]
          }
        : false,
    middleware: getDefaultMiddleware => {
      const middlewares = getDefaultMiddleware({
        immutableCheck:
          IS_NODE_ENV_DEVELOPMENT &&
          process.env.GMS_DISABLE_REDUX_IMMUTABLE_CHECK?.toLocaleLowerCase() !== 'true',
        serializableCheck:
          IS_NODE_ENV_DEVELOPMENT &&
          process.env.GMS_DISABLE_REDUX_SERIALIZABLE_CHECK?.toLocaleLowerCase() !== 'true'
      })
        .concat(thunkMiddleware)
        .concat(systemEventGatewayApiSlice.middleware)
        .concat(eventManagerApiSlice.middleware)
        .concat(processingConfigurationApiSlice.middleware)
        .concat(processingStationApiSlice.middleware)
        .concat(signalEnhancementConfigurationApiSlice.middleware)
        .concat(sohAceiApiSlice.middleware)
        .concat(ssamControlApiSlice.middleware)
        .concat(stationDefinitionSlice.middleware)
        .concat(systemMessageDefinitionApiSlice.middleware)
        .concat(userManagerApiSlice.middleware)
        .concat(workflowApiSlice.middleware);

      if (!DISABLE_REDUX_STATE_SYNC) {
        middlewares.push(createStateSyncMiddleware());
      }

      // ! the logger should always be the last middleware added
      // enable the Redux logger only if `GMS_ENABLE_REDUX_LOGGER` is set to true
      if (process.env.GMS_ENABLE_REDUX_LOGGER?.toLocaleLowerCase() === 'true') {
        const reduxLogger = createLogger({
          collapsed: true,
          duration: true,
          timestamp: false,
          level: 'info',
          logger: console,
          logErrors: true,
          diff: false
        });
        middlewares.push(reduxLogger);
      }

      return middlewares;
    },
    enhancers:
      electron && electronEnhancer
        ? [
            // must be placed after the enhancers which dispatch
            // their own actions such as redux-thunk or redux-saga
            electronEnhancer({
              dispatchProxy: a => store.dispatch(a)
            })
          ]
        : []
  });

  // initialize state using any previous state from other tabs
  if (!DISABLE_REDUX_STATE_SYNC) {
    initStateWithPrevTab(store);
  }

  return store;
};

export interface GMSWindow extends Window {
  ReduxStore: ReturnType<typeof buildStore> | undefined;
}

const gmsWindow = (window as unknown) as GMSWindow;

// ! ensure that only one Redux store instance is created
if (gmsWindow.ReduxStore === undefined) {
  // !ensure a single instance of the Redux store by attaching it to the window object
  // ?this also exposes the Redux store for Cypress testing
  gmsWindow.ReduxStore = buildStore();

  // store is created connect to System Event gateway and set callback to set connection status
  establishWsConnection();
  const updateConnectionStatus = (connected: boolean): void => {
    if (gmsWindow.ReduxStore) {
      gmsWindow.ReduxStore.dispatch(userSessionActions.setConnected(connected));
    }
  };

  registerConnectionStatusCallback(
    updateConnectionStatus,
    gmsWindow.ReduxStore.getState().app.userSession
  );
} else {
  // !Throw an error if the Redux store is created more than once
  const error = new Error('Critical Error: Redux store has already been created');
  logger.error('Redux store should only be created once', error);
  throw error;
}

/**
 * Creates the Redux application store, which is stored on the window object in order to avoid any chance for
 * multiple stores to be created (if multiple copies of the module are included in the bundle, for example).
 *
 * @returns the Redux store for the application
 */
export const getStore = (): typeof gmsWindow.ReduxStore => gmsWindow.ReduxStore;

/**
 * The application state.
 * Infer the `AppState` types from the store itself
 */
export type AppState = ReturnType<typeof reducer>;

/**
 * The application dispatched (typed).
 * Infer the `AppDispatch` types from the store itself
 */
export type AppDispatch = typeof gmsWindow.ReduxStore.dispatch;
