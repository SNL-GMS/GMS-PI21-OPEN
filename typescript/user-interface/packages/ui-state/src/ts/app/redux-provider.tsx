import React from 'react';
import { Provider } from 'react-redux';

import { getStore } from './store';
/**
 * Wraps the provided component with Redux Provider.
 *
 * @param Component the component
 * @param store the redux store
 */
export const withReduxProvider = (
  Component: React.ComponentClass | React.FunctionComponent
): React.ComponentClass =>
  class WithReduxProvider extends React.PureComponent {
    /**
     * Wrap the component in a redux providers
     */
    public render(): JSX.Element {
      return (
        <Provider store={getStore()}>
          {/* eslint-disable-next-line react/jsx-props-no-spreading */}
          <Component {...this.props} />
        </Provider>
      );
    }
  };
