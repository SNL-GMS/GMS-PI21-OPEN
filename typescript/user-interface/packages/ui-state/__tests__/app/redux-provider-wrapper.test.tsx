import { H1 } from '@blueprintjs/core';
import type { ReactWrapper } from 'enzyme';
import React from 'react';

import { withReduxProvider } from '../../src/ts/app/redux-provider';

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
const Enzyme = require('enzyme');

window.alert = jest.fn();
window.open = jest.fn();

// simple component we can wrap
class Welcome extends React.PureComponent {
  public render() {
    return <H1>Hello</H1>;
  }
}

describe('Redux wrapper', () => {
  const component: any = Welcome;
  const Wrapper = withReduxProvider(component);

  // make sure the function is defined
  test('should exist', () => {
    expect(withReduxProvider).toBeDefined();
  });

  // see what we got from the wrapper (should be a constructor function for a class)
  test('function should create a component class', () => {
    // returns a class function that we can call with the new keyword
    expect(typeof Wrapper).toBe('function');
  });

  // TODO Fix this tests nothing
  // lets render our wrapper and see what we get back
  test.skip('can create a rendered wrapper', () => {
    const mountedWrapper: ReactWrapper = Enzyme.mount(<Wrapper />);
    // make sure we can make a client if we call the client creation function
    expect(mountedWrapper).toMatchSnapshot();
  });
});
