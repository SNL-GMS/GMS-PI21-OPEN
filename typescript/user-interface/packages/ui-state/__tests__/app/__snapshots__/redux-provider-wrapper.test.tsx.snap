// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux wrapper can create a rendered wrapper 1`] = `
<div>
  <h1
    class="bp4-heading"
  >
    Hello
  </h1>
</div>
`;
